### ISSUE TITLE
[RFC][HttpClient] Add $response->wait();

### ISSUE THREAD

**Issue by lyrixx** [symfony/symfony#43410]
**Description**

Sometime, you want to perform HTTP request, but you don't care about the content **but** you want the request to finish before continuing to do something.
For example, when you call the "refresh" endpoint  on Elasticsearch. You may want to be sure the refresh is done before doing anything else.

ATM, to do so you have two options:

* `$response->getContent(false)`:
  1. It creates a buffer for nothing.
  2. It creates WTF during code review and lead to question like "Why are you calling this method whereas you don't use the content (true story)
* `foreach ($client->stream($response) as $chunk)`:
  1. It's for people that know internals
  2. the core is a bit cryptic, it does not tell what it does.
  3. oneliner, without `{}` , some people don't like (note: I don't care, but still)

So, basically, with both options, a comment should be added to the code to tell what it does.

I'm not happy with theses solutions, that's why I propose another solution:

```php
$response->wait();
```

1. that's simple
2. that's easy
3. that's explicit

**Example**
```php
$response = $this->esClient->request('POST', '/index/_docs', [...]);
// check if response is OK
$response = $this->esClient->request('POST', '/index/_refresh');
$response->wait(throw: true);
```

**Comment by stof** [comment]
makes sense to me. /cc @nicolas-grekas

**Comment by nicolas-grekas** [comment]
Adding the method to the component would break the abstraction provided by the contracts.
And adding the method to the contracts would 1. require a BC break 2. provide an alternative way to do something that is already possible using a more powerful method.
Calling `$client->stream()` is not internal at all.
My reco here would be to create this `wait()` method in userland. That would fix the issue with reviewers.
I also don't think this need is frequent enough to deserve a dedicated method (waiting for the content and trashing it).
TL;DR, I'm not sold at all :)

**Comment by stof** [comment]
@nicolas-grekas maybe symfony/http-client could expose some high-level helpers, allowing to make the intent more clear for cases like that by having an actual `wait($response)` method, without the need to make it a method on the response (and so a change in the contract).

`$client->stream()` is not internal, but it is definitely low-level.

**Comment by lyrixx** [comment]
Yeah, when I said `internal` I meant `low-level`

BC, there is not BC break, we can provide a BC path. But we can also add an helper. I don't care on this point

**Comment by lyrixx** [comment]
> I also don't think this need is frequent enough to deserve a dedicated method (waiting for the content and trashing it).

Side note: **MANY** people think the following code will wait for the response:

```
$response = $this->esClient->request('POST', '/index/_docs', [...]);
or 
$this->esClient->request('POST', '/index/_docs', [...]);
```

So people are not aware that the client is async by default. I answered many question / help teams during coaching sessions about this very topic.

So I do think this need is frequent, but people don't realize they need it because they think it's done by default.

> I don't understand, my ES index is not up to date when I post some data, but when I refresh the page it's OK

Is a very common issue!

**Comment by derrabus** [comment]
> ```php
> $this->esClient->request('POST', '/index/_docs', [...]);
> ```

I would indeed expect that this call will wait for the response because the returned response object is destructed immediantly and its destructor is blocking.

Test script:

```php
$microtime = microtime(true);

$httpClient = HttpClient::create();
$httpClient->request('POST', 'https://httpbin.org/delay/5');

dump(microtime(true) - $microtime);
```

**Comment by nicolas-grekas** [comment]
The destructor waits for the headers only. This is by design. If the code doesn't care about the body, neither does the destructor.

**Comment by lyrixx** [comment]
Wep, this is misleading.

That's why I do think a `$response->wait()` method could be useful

**Comment by nicolas-grekas** [comment]
Another argument against wait is async: calling it on responses forces sync operations.
At least it should be `$client->wait($responses|$response)`. But I'm still not sold this should be part of the abstraction, since the abstraction already covers that need.

This is the only code you need to add in userland to cover this:
```php
private function wait(iterable|ResponseInterface $responses)
{
	foreach ($this->client->stream($responses) as $response => $chunk);
}
```

**Comment by lyrixx** [comment]
Yes it breaks async but that's the point 🤔

In my class I have different clients ( with different timeout configuration) so it Would be harder that your snippet.

But still, it's not clear for people (even in the core team), so I'm still thinking it's a good idea

**Comment by derrabus** [comment]
> The destructor waits for the headers only. This is by design. If the code doesn't care about the body, neither does the destructor.

Right, but I assumed that this is what @lyrixx wanted when he wrote…

> Sometime, you want to perform HTTP request, but you don't care about the content

When I don't care about the response body, waiting for the headers only (to see if my request has been successful) sounds sufficient to me. What would be the use case of waiting until the whole body is streamed if I intend to scrap it anyway?

**Comment by lyrixx** [comment]
When ES index some content for example.

**Comment by nicolas-grekas** [comment]
> When ES index some content for example.

I understand this works in practice, but this looks like abusing HTTP. It'd be nice if ES would return a series of `102 Processing` instead. I wouldn't add a mainstream method to support that...

**Comment by lyrixx** [comment]
Hmm, indeed I understand your POV.

And indeed, it's very strange to return a 200 but continuing to processing something in the background.

I'm closing this issue for the reason you just gave.
